### ISSUE TITLE
Global shortcuts

### ISSUE THREAD

**Issue by wanderingmatt** [codetree/feedback#48]
I would like to be able to specify a Shortcut as "global" so that every member of the organization can see it. We started testing out Codetree for review during bug meetings and, depending on who leads the meeting, we can't rely on the same shortcuts for review.

**Comment by derrickreimer** [comment]
:+1:

**Comment by dblandin** [comment]
:+1:

I currently have two shortcuts setup that show open pull requests with either the label `review` or `qa`. I would also love the ability to promote these personal shortcuts to the organization level.

**Comment by dblandin** [comment]
I noticed these new global shortcuts based on my custom stages show up in the sidebar:

![screen shot 2015-02-11 at 11 28 00 am](https://cloud.githubusercontent.com/assets/123595/6155553/cffab886-b1f9-11e4-9eca-77c10df9fa53.png)

Great to see them, but can we widen the scope to Issues and Pull Requests? We generally use the custom stages for PRs, not issues.

**Comment by derrickreimer** [comment]
@dblandin would you like your default view to always show issues and pull requests together, or just for these custom stages?

**Comment by wanderingmatt** [comment]
FWIW, my team considers Pull Requests just another stage of an issue.

**Comment by dblandin** [comment]
@WanderingMatt Yep, our team as well.

@djreimer It would be great to see the counts of open PRs next to the global shortcuts for custom stages so I can see at a glance: 2 PRs are waiting on QA, 3 PRs need some code review, etc...

I'm fine with the current default view behavior which takes you to open issues with a separate link for open PRs:

![screen shot 2015-02-12 at 5 35 32 pm](https://cloud.githubusercontent.com/assets/123595/6179297/7c9b7fec-b2dd-11e4-813d-69fd2fa96df4.png)

**Comment by derrickreimer** [comment]
Thanks for the info guys.

Here's the workflow I've been recommending:
1. File standard issues for bugs and features
2. Attach pull requests to issues using the GitHub reference convention (e.g. fixes #12345)
3. When it's ready for review/QA, move the original issue to that stage
4. Once the pull request is merged/deployed/QA'd, then close the original issue

There are several advantages that I've found with this approach:
- Some issues don't go through a pull request phase, depending on the nature of the issue
- Some issues require multiple pull requests, so tracking overall state with a parent issue keeps things organized
- Since I tend to have a 1-1 correspondence between issues and pull requests, seeing them together generally shows me two records per bug/feature == more noise to parse through.

Questions for you @WanderingMatt @dblandin:
- Do you tend to have a parent issue that you connect your PRs to?
- If yes, what do you do with the parent issue vs. the PR on your task board as you make progress on the issue? (e.g. leave the original issue in "To Do"? or move the issue and the PR through stages simultaneously?)

**Comment by dblandin** [comment]
I do like keeping the lists of issues and PRs separate. We would also have some visual redundancy if both were shown by default. It may be related to our specific stages (I'd love to hear other examples) but our issues aren't involved in custom stages as PRs are. Issues aren't code reviewed, PRs are. Issues aren't QA'd, changes introduced in PRs are.

To address your first question: generally yes, we tend to have a parent issue for any PR we open. We usually reference the parent issue when opening the PR with a

> Addresses #some-issue-number

And the second: we usually at least assign the parent issue to mark that it's currently being addressed and is "in progress".

The PR will then go through stages of dev, discussion, code review, and QA. Once a PR passes QA, it's merged and closed. The parent issue, if satisfied at that point, is closed as well with a:

> Closing with #some-pr-number

Thanks for discussing this! Excited to see where this leads. :)

**Comment by dblandin** [comment]
@djreimer Hey, that was quick!! This is perfect! :+1:

![screen shot 2015-02-13 at 2 01 23 pm](https://cloud.githubusercontent.com/assets/123595/6194318/e1ac1a4e-b388-11e4-9d0e-8ab6dd20444f.png)

**Comment by dblandin** [comment]
Ha, never mind. That's my own shortcut... whoops :persevere:

**Comment by wanderingmatt** [comment]
To try and answer your first question — our PRs have either a 1-to-1 or 1-to-many relationship with our issues. We file issues for every bug/feature. If it is a one-off maintenance bug, we create a matching branch and (eventually) a matching PR for that branch. For feature work however, we create an omnibus branch which we will hang the eventual PR off of. As issues get worked on, we create matching branches which then get merged into the omnibus branch.

Currently, we create PRs when either of these situations reaches the review stage; however, we're considering bumping PR creation down to the "In Process" stage because we find the code comparison and discussion they offer to be valuable as soon as anyone starts working on anything.

Regarding your second question, we don't do anything with the issues right now, but it would be great if the issues "label" tracked it's parent PR.

All things considered, I think if you added the ability to choose type as a criteria for the task board stages (I previously filed #47 for this), that would solve the issue for our team that @dblandin brought up in this issue. That way, when we're looking at the task board for a given milestone, the "Backlog" through "In Process" stages would be filled with issues, whereas, the "Needs Review" through "Done" stages would show PRs. Hope that makes sense.

It would be icing on the cake if you considered what you mentioned above — having issue stage labels track that of their parent PR. That way, as we moved our PRs around on the task board, the corresponding issues labels would match in case someone were to be looking at them from the Github side.

**Comment by derrickreimer** [comment]
Alright, here's the latest:
- Custom stages now include issues and pull requests (on boards and in the left-hand shortcuts), as requested by @dblandin
- Pull requests attached to issues via the "fixes" convention now show on task board cards

Still to come:
- Pull requests listed under issues on boards will be draggable to different stages. E.g. if an issue is "in progress" and you want it's attached PR to enter the "needs review" stage, you will be able to grab the PR on the parent issue and drag to a different column:

![drag-prs](https://cloud.githubusercontent.com/assets/341387/6263064/adc0f2f4-b7c5-11e4-881e-090cf1444207.png)

I'm really stoked about this latest iteration and will hopefully have this last part done in the next day or so. Thanks for all the input here! :tada:

**Comment by dblandin** [comment]
Sounds like an awesome update! Thanks for your hard work on this!

Will PRs attach to issues on the task board via any other mention keyword besides "fixes"? We usually use "fixes" when a PR should close a bug report issue and "addresses" for anything else (like a feature request).

**Comment by derrickreimer** [comment]
> Will PRs attach to issues on the task board via any other mention keyword besides "fixes"? We usually use "fixes" when a PR should close a bug report issue and "addresses" for anything else (like a feature request).

I like that convention. I think it would be reasonable to expand this to list any PRs referencing an issue. Stay tuned.

**Comment by wanderingmatt** [comment]
> I think it would be reasonable to expand this to list any PRs referencing an issue.

:+1:

**Comment by derrickreimer** [comment]
Alright, all referencing PRs should be visible now.

**Comment by dblandin** [comment]
Fantastic! Thanks @djreimer!

**Comment by derrickreimer** [comment]
:metal:

**Comment by derrickreimer** [comment]
Shipped! :shipit: 

![global-shortcuts](https://cloud.githubusercontent.com/assets/341387/10712776/70980f7e-7a58-11e5-97ed-9049f9d35db2.png)
